### PR TITLE
use VertexArrayObject extension when available

### LINKED ISSUE #1361: Use OES_vertex_array_object API when available
https://www.khronos.org/registry/webgl/extensions/OES_vertex_array_object/

Per @ansis' comment at https://github.com/mapbox/mapbox-gl-js/pull/1356#issuecomment-118356001

Related to #1318 #1356

### PR DESCRIPTION
**Pull request by ansis**
fixes #1361

When the extension is available, use it to record and re-apply buffer binds and attrib pointer setting. When the extension isn't available this just sets the pointers and binds the buffers each time.

This also namespaces bucket buffers/arrays with objects instead of string concatenation.

I ran the frame-duration benchmark four times for master (3a95d00) and this branch:

```
vao:        15.3 ms
master:     17.9 ms
vao:        15.5 ms
master:     18.2 ms
vao:        14.2 ms
master:     18.5 ms
vao:        15.2 ms
master:     17.4 ms

vao average:    15 ms
master average: 18 ms
```

This branch takes **83%** of the time that master takes.

:eyes: @lucaswoj

### CODE REVIEW

File: js/data/bucket.js [hunk 1]
```diff
@@ -76,10 +77,33 @@ function Bucket(options) {
...
     if (options.elementGroups) {
         this.elementGroups = options.elementGroups;
-        this.buffers = util.mapObject(options.arrays, function(array, bufferName) {
-            var arrayType = options.arrayTypes[bufferName];
-            var type = (arrayType.members.length && arrayType.members[0].name === 'vertices' ? Buffer.BufferType.ELEMENT : Buffer.BufferType.VERTEX);
-            return new Buffer(array, arrayType, type);
+
+        for (var programName in this.elementGroups) {
+            var elementGroups = this.elementGroups[programName];
+            if (elementGroups) {
```
Comment (lucaswoj) [inline js/data/bucket.js]: I'm not thrilled with the idea of using `if (elementGroups)` as a way to check if this is running on the main thread or in the worker. However, I appreciate why this is really hard given the current architecture of the `Bucket` class. Whenever you are done making changes to this part of the system, I would like to refactor `Bucket` 👍

File: js/render/painter.js [hunk 2]
```diff
@@ -7,31 +7,21 @@ var TilePyramid = require('../source/tile_pyramid');
...
-
 /**
  * Initialize a new painter object.
  *
  * @param {Canvas} gl an experimental-webgl drawing context
  * @private
  */
 function Painter(gl, transform) {
-    this.gl = util.extend(gl, glPolyfill);
+    this.gl = gl;
```
Comment (lucaswoj) [inline js/render/painter.js]: 🎉

**Comment by lucaswoj** [comment]
This looks great! Ready to 🚢 as is.

File: js/render/vertex_array_object.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    this.boundVertexBuffer = null;
+    this.boundVertexBuffer2 = null;
+    this.boundVertexOffset = null;
+    this.boundElementBuffer = null;
+    this.vao = null;
+}
+
+var reported = false;
+
+VertexArrayObject.prototype.bind = function(gl, program, vertexBuffer, vertexBuffer2, vertexOffset, elementBuffer) {
```
Comment (lucaswoj) [inline js/render/vertex_array_object.js]: Can we be consistent in referring to this as either `vertexBuffer2` or `secondVertexBuffer`? (I'm partial to `vertexBuffer2` but not strongly opinionated on this issue.)

Comment (ansis) [inline js/render/vertex_array_object.js]: I'll add a switch from `secondElement` to `element2` in https://github.com/mapbox/mapbox-gl-js/pull/2488/files

**Pull request merged**